File: PHRPReader/clsSearchEngineParameters.vb
Option Strict On

Public Class clsSearchEngineParameters

    Public Const MASS_TYPE_MONOISOTOPIC As String = "monoisotopic"
    Public Const MASS_TYPE_AVERAGE As String = "average"

    Private mSearchEngineName As String
    Private mSearchEngineVersion As String
    Private mSearchDate As DateTime

    Private mFastaFilePath As String

    Private mPrecursorMassToleranceDa As Double     ' Precursor mass tolerance, in Da; 0 if unknown
    Private mPrecursorMassTolerancePpm As Double    ' Precursor mass tolerance, in ppm; 0 if unknown

    Private mPrecursorMassType As String            ' Typically "monoisotopic" or "average"
    Private mFragmentMassType As String

    Private mEnzyme As String
    Private mMaxNumberInternalCleavages As Integer
    Private mMinNumberTermini As Integer                        ' 0 means no-enzyme, 1 means partially tryptic, 2 means fully tryptic

    Private ReadOnly mModInfo As List(Of clsModificationDefinition)

    Private ReadOnly mParameters As Dictionary(Of String, String)

#Region "Properties"
    Public Property Enzyme As String
        Get
            Return mEnzyme
        End Get
        Set(value As String)
            If value = "" Then value = "none"
            mEnzyme = value
        End Set
    End Property

    Public Property FastaFilePath As String
        Get
            Return mFastaFilePath
        End Get
        Set(value As String)
            mFastaFilePath = value
        End Set
    End Property

    Public Property FragmentMassType As String
        Get
            Return mFragmentMassType
        End Get
        Set(value As String)
            If String.IsNullOrEmpty(value) Then value = MASS_TYPE_MONOISOTOPIC
            mFragmentMassType = value
        End Set
    End Property

    Public Property MaxNumberInternalCleavages As Integer
        Get
            Return mMaxNumberInternalCleavages
        End Get
        Set(value As Integer)
            mMaxNumberInternalCleavages = value
        End Set
    End Property

    Public Property MinNumberTermini As Integer
        Get
            Return mMinNumberTermini
        End Get
        Set(value As Integer)
            mMinNumberTermini = value
        End Set
    End Property

    Public ReadOnly Property ModInfo As List(Of clsModificationDefinition)
        Get
            Return mModInfo
        End Get
    End Property

    Public ReadOnly Property Parameters As Dictionary(Of String, String)
        Get
            Return mParameters
        End Get
    End Property

    Public Property PrecursorMassToleranceDa As Double
        Get
            Return mPrecursorMassToleranceDa
        End Get
        Set(value As Double)
            mPrecursorMassToleranceDa = value
        End Set
    End Property

    Public Property PrecursorMassTolerancePpm As Double
        Get
            Return mPrecursorMassTolerancePpm
        End Get
        Set(value As Double)
            mPrecursorMassTolerancePpm = value
        End Set
    End Property

    Public Property PrecursorMassType As String
        Get
            Return mPrecursorMassType
        End Get
        Set(value As String)
            If String.IsNullOrEmpty(value) Then value = MASS_TYPE_MONOISOTOPIC
            mPrecursorMassType = value
        End Set
    End Property

    Public ReadOnly Property SearchEngineName As String
        Get
            Return mSearchEngineName
        End Get
    End Property

    Public ReadOnly Property SearchEngineVersion As String
        Get
            Return mSearchEngineVersion
        End Get
    End Property

    Public ReadOnly Property SearchDate As DateTime
        Get
            Return mSearchDate
        End Get
    End Property

#End Region

    Public Sub New(searchEngineName As String)
        Me.New(searchEngineName, New List(Of clsModificationDefinition), Nothing)
    End Sub

    Public Sub New(searchEngineName As String, objModInfo As List(Of clsModificationDefinition))
        Me.New(searchEngineName, objModInfo, Nothing)
    End Sub

    Public Sub New(searchEngineName As String, objModInfo As List(Of clsModificationDefinition), Parameters As Dictionary(Of String, String))
        Me.InitializeDefaults()

        mSearchEngineName = searchEngineName

        mModInfo = objModInfo

        If objModInfo Is Nothing Then
            mModInfo = New List(Of clsModificationDefinition)
        Else
            mModInfo = objModInfo
        End If

        If mParameters Is Nothing Then
            mParameters = New Dictionary(Of String, String)(StringComparer.CurrentCultureIgnoreCase)
        Else
            mParameters = Parameters
        End If

    End Sub

    Public Sub AddModification(objModInfo As clsModificationDefinition)
        mModInfo.Add(objModInfo)
    End Sub

    Public Sub AddUpdateParameter(kvSetting As KeyValuePair(Of String, String))
        AddUpdateParameter(kvSetting.Key, kvSetting.Value)
    End Sub

    Public Sub AddUpdateParameter(ParamName As String, ParamValue As String)
        If mParameters.ContainsKey(ParamName) Then
            mParameters(ParamName) = ParamValue
        Else
            mParameters.Add(ParamName, ParamValue)
        End If
    End Sub

    Public Sub ClearModifications()
        mModInfo.Clear()
    End Sub

    Public Sub ClearParameters()
        mParameters.Clear()
    End Sub

    Private Sub InitializeDefaults()
        mSearchEngineName = "Unknown"
        mSearchEngineVersion = "Unknown"
        mSearchDate = New DateTime(1980, 1, 1)

        mFastaFilePath = String.Empty

        mPrecursorMassToleranceDa = 0
        mPrecursorMassTolerancePpm = 0

        mPrecursorMassType = MASS_TYPE_MONOISOTOPIC
        mFragmentMassType = MASS_TYPE_MONOISOTOPIC

        mEnzyme = "trypsin"
        mMaxNumberInternalCleavages = 4
        mMinNumberTermini = 0

    End Sub

    Public Sub UpdateSearchEngineVersion(strSearchEngineVersion As String)
        mSearchEngineVersion = String.Copy(strSearchEngineVersion)
    End Sub

    Public Sub UpdateSearchDate(dtSearchDate As DateTime)
        mSearchDate = dtSearchDate
    End Sub

End Class

File: PHRPReader/clsSearchEngineParameters.cs
using System;
using System.Collections.Generic;

namespace PHRPReader
{
    /// <summary>
    /// Search engine parameters container
    /// </summary>
    public class clsSearchEngineParameters
    {
        /// <summary>
        /// Monoisotopic mass
        /// </summary>
        public const string MASS_TYPE_MONOISOTOPIC = "monoisotopic";

        /// <summary>
        /// Average mass
        /// </summary>
        public const string MASS_TYPE_AVERAGE = "average";

        private string mSearchEngineName;
        private string mSearchEngineVersion;
        private DateTime mSearchDate;

        private string mPrecursorMassType;             // Typically "monoisotopic" or "average"
        private string mFragmentMassType;

        private string mSearchEngineParamFilePath;

        private string mEnzyme;

        private readonly List<clsModificationDefinition> mModInfo;

        private readonly Dictionary<string, string> mParameters;

        #region "Properties"
        /// <summary>
        /// Enzyme name
        /// </summary>
        public string Enzyme
        {
            get => mEnzyme;
            set
            {
                if (string.IsNullOrEmpty(value))
                    value = "none";
                mEnzyme = value;
            }
        }

        /// <summary>
        /// FASTA file path
        /// </summary>
        public string FastaFilePath { get; set; }

        /// <summary>
        /// Fragment mass type
        /// </summary>
        /// <remarks>Typically "monoisotopic" or "average"</remarks>
        public string FragmentMassType
        {
            get => mFragmentMassType;
            set
            {
                if (string.IsNullOrEmpty(value))
                    value = MASS_TYPE_MONOISOTOPIC;
                mFragmentMassType = value;
            }
        }

        /// <summary>
        /// Maximum number of internal cleavages (missed cleavage points)
        /// </summary>
        public int MaxNumberInternalCleavages { get; set; }

        /// <summary>
        /// 0 means no-enzyme, 1 means partially tryptic, 2 means fully tryptic
        /// </summary>
        /// <remarks>For trypsin, this is NTT or Number of Tryptic Termini</remarks>
        public int MinNumberTermini { get; set; }

        /// <summary>
        /// Dynamic and static mods to search for
        /// </summary>
        public List<clsModificationDefinition> ModInfo => mModInfo;

        /// <summary>
        /// Parameter dictionary (key/value pairs)
        /// </summary>
        public Dictionary<string, string> Parameters => mParameters;

        /// <summary>
        /// Precursor mass tolerance, in Daltons; 0 if unknown
        /// </summary>
        public double PrecursorMassToleranceDa { get; set; }

        /// <summary>
        /// Precursor mass tolerance, in ppm; 0 if unknown
        /// </summary>
        public double PrecursorMassTolerancePpm { get; set; }

        /// <summary>
        /// Precursor mass type
        /// </summary>
        /// <remarks>Typically "monoisotopic" or "average"</remarks>
        public string PrecursorMassType
        {
            get => mPrecursorMassType;
            set
            {
                if (string.IsNullOrWhiteSpace(value))
                    value = MASS_TYPE_MONOISOTOPIC;
                mPrecursorMassType = value;
            }
        }

        /// <summary>
        /// Search engine name
        /// </summary>
        public string SearchEngineName => mSearchEngineName;

        /// <summary>
        /// Search engine parameter file path
        /// </summary>
        public string SearchEngineParamFilePath => string.IsNullOrWhiteSpace(mSearchEngineParamFilePath) ? string.Empty : mSearchEngineParamFilePath;

        /// <summary>
        /// Search engine version
        /// </summary>
        public string SearchEngineVersion => mSearchEngineVersion;

        /// <summary>
        /// Search date
        /// </summary>
        public DateTime SearchDate => mSearchDate;

        #endregion

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="searchEngineName"></param>
        public clsSearchEngineParameters(string searchEngineName) : this(searchEngineName, new List<clsModificationDefinition>(), null)
        {
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="searchEngineName"></param>
        /// <param name="modInfo"></param>
        public clsSearchEngineParameters(string searchEngineName, List<clsModificationDefinition> modInfo) : this(searchEngineName, modInfo, null)
        {
        }

        /// <summary>
        /// Constructor
        /// </summary>
        /// <param name="searchEngineName"></param>
        /// <param name="modInfo"></param>
        /// <param name="parameters"></param>
        public clsSearchEngineParameters(string searchEngineName, List<clsModificationDefinition> modInfo, Dictionary<string, string> parameters)
        {
            InitializeDefaults();

            mSearchEngineName = searchEngineName;
            mSearchEngineParamFilePath = string.Empty;

            mModInfo = modInfo;

            mModInfo = modInfo ?? new List<clsModificationDefinition>();

            if (mParameters == null)
            {
                mParameters = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
            }
            else
            {
                mParameters = parameters;
            }
        }

        /// <summary>
        /// Add a new dynamic or static modification
        /// </summary>
        /// <param name="modInfo"></param>
        public void AddModification(clsModificationDefinition modInfo)
        {
            mModInfo.Add(modInfo);
        }

        /// <summary>
        /// Add/update a parameter
        /// </summary>
        /// <param name="kvSetting"></param>
        public void AddUpdateParameter(KeyValuePair<string, string> kvSetting)
        {
            AddUpdateParameter(kvSetting.Key, kvSetting.Value);
        }

        /// <summary>
        /// Add/update a parameter
        /// </summary>
        /// <param name="paramName"></param>
        /// <param name="paramValue"></param>
        public void AddUpdateParameter(string paramName, string paramValue)
        {
            if (mParameters.ContainsKey(paramName))
            {
                mParameters[paramName] = paramValue;
            }
            else
            {
                mParameters.Add(paramName, paramValue);
            }
        }

        /// <summary>
        /// Clear stored dynamic and static modifications
        /// </summary>
        public void ClearModifications()
        {
            mModInfo.Clear();
        }

        /// <summary>
        /// Clear stored key/value parameters
        /// </summary>
        public void ClearParameters()
        {
            mParameters.Clear();
        }

        private void InitializeDefaults()
        {
            mSearchEngineName = "Unknown";
            mSearchEngineVersion = "Unknown";
            mSearchDate = new DateTime(1980, 1, 1);

            FastaFilePath = string.Empty;

            PrecursorMassToleranceDa = 0;
            PrecursorMassTolerancePpm = 0;

            mPrecursorMassType = MASS_TYPE_MONOISOTOPIC;
            mFragmentMassType = MASS_TYPE_MONOISOTOPIC;

            mEnzyme = "trypsin";
            MaxNumberInternalCleavages = 4;
            MinNumberTermini = 0;
        }

        /// <summary>
        /// Update the search engine parameter file path
        /// </summary>
        /// <param name="paramFilePath"></param>
        public void UpdateSearchEngineParamFilePath(string paramFilePath)
        {
            mSearchEngineParamFilePath = paramFilePath;
        }

        /// <summary>
        /// Update the search engine version
        /// </summary>
        /// <param name="searchEngineVersion"></param>
        public void UpdateSearchEngineVersion(string searchEngineVersion)
        {
            mSearchEngineVersion = string.Copy(searchEngineVersion);
        }

        /// <summary>
        /// Update the search date
        /// </summary>
        /// <param name="searchDate"></param>
        public void UpdateSearchDate(DateTime searchDate)
        {
            mSearchDate = searchDate;
        }
    }
}
